File: Jump_Game_II_v1.kt
import kotlin.math.min

class Solution {
    fun jump(nums: IntArray): Int {
        if (nums.isEmpty() || nums.size == 1) return 0
        val minimumNumberOfJumps = IntArray(nums.size) { 0xffffff }
        minimumNumberOfJumps[0] = 0
        var longest = 0
        for ((i, j) in nums.withIndex()) {
            if (i + j <= longest) continue
            longest = i + j
            if (longest >= nums.lastIndex) return minimumNumberOfJumps[i] + 1
            for (k in 1..j) {
                if (i + k >= nums.size) break
                minimumNumberOfJumps[i + k] = min(minimumNumberOfJumps[i + k], minimumNumberOfJumps[i] + 1)
            }
        }
        return minimumNumberOfJumps.last()
    }
}

fun main() {
    val solution = Solution()
    // 2 2 0 2
    val testCases = arrayOf(intArrayOf(2, 3, 1, 1, 4), intArrayOf(2, 3, 0, 1, 4), intArrayOf(1), intArrayOf(7, 0, 9, 6, 9, 6, 1, 7, 9, 0, 1, 2, 9, 0, 3))
    for (case in testCases) {
        println(solution.jump(case))
    }
}

File: Jump_Game_II_v1.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        if (nums.size() <= 1) {
            return 0;
        }
        vector<int> minSteps(nums.size(), INT_MAX);
        minSteps[0] = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (minSteps[i] == INT_MAX) {
                break;
            }
            for (int j = 1; j <= nums[i] && i + j < nums.size(); j++) {
                minSteps[i + j] = min(minSteps[i + j], minSteps[i] + 1);
            }
        }
        if (minSteps[nums.size() - 1] == INT_MAX) {
            return -1;
        }
        return minSteps[nums.size() - 1];
    }
};

int main() {
    Solution sol;
    vector<vector<int>> testCases = {{2, 3, 1, 1, 4}, {3, 2, 1, 0, 4}, {2, 1, 2, 2, 1, 2, 2, 2}};
    for (auto testCase : testCases) {
        cout << sol.jump(testCase) << endl;
    }
    return 0;
}